## Vishal Patel 
#### pandas-challenge

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [3]:
total_players = len(purchase_data['SN'].value_counts())
print(f"Total number of unique players: {total_players}")

Total number of unique players: 576


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
#I put alot of outputs manually in a dataframe to see clearly from here on out
unique_items = len(purchase_data['Item ID'].value_counts())
average_price = '${:,.2f}'.format(purchase_data['Price'].mean())
num_purchases = purchase_data.shape[0]
total_rev = '${:,.2f}'.format(purchase_data['Price'].sum())
analysis_df = pd.DataFrame({
    "Number of Unique Items":[unique_items],
    "Average Price":[average_price],
    "Number of Purchases":[num_purchases],
    "Total Revenue": [total_rev]
})
analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [5]:
gender_df = purchase_data.groupby("Gender")
gender_dfcounts = gender_df["SN"].nunique()
gender_dfpercentages = gender_df["SN"].nunique()/float(total_players)
genderdemodf = pd.DataFrame({"Total Count": gender_dfcounts,
                              "Percentage of Players": gender_dfpercentages})
genderdemodf["Percentage of Players"] = genderdemodf["Percentage of Players"].map("{:.2%}".format)

#This below is just formatting it to get it to look like the exmaple output given. The above code gives
#Female then Male then Other, and the code below switched Female and Male so that it looks like the output
#I don't believe this is neccessary but it is for grading purposes just in case
b, c = genderdemodf.iloc[0].copy(), genderdemodf.iloc[1].copy()
genderdemodf.iloc[0],genderdemodf.iloc[1] = c,b
genderdemodf = genderdemodf.rename(index = {"Female": "Male", "Male": "Female"})
genderdemodf

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [6]:
#I used gender_df from above, it is the grouped by Gender
gendercount = gender_df["SN"].count()
genderavg = gender_df["Price"].mean()
gendertotal = gender_df["Price"].sum()
genderavgtotal = gender_df["Price"].sum()/gender_df["SN"].nunique()
genderpuranalysis_df = pd.DataFrame({"Purchase Count": gendercount,
                              "Average Purchase Price": genderavg,
                              "Total Purchase Value": gendertotal,
                              "Avg Total Purchase per Person": genderavgtotal})
genderpuranalysis_df["Average Purchase Price"] = genderpuranalysis_df["Average Purchase Price"].map("${:.2f}".format)
genderpuranalysis_df["Total Purchase Value"] = genderpuranalysis_df["Total Purchase Value"].map("${:.2f}".format)
genderpuranalysis_df["Avg Total Purchase per Person"] = genderpuranalysis_df["Avg Total Purchase per Person"].map("${:.2f}".format)
genderpuranalysis_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [7]:
bins = [0, 9.99, 14, 19, 24, 29, 34, 39, 100]
groupnames = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

purchase_data["Age Demographics"] = pd.cut(purchase_data["Age"], bins, labels = groupnames, include_lowest = True)

groupeddemo = purchase_data.groupby("Age Demographics")

total_counts = list(groupeddemo["SN"].nunique())

percentages = ["{:.2%}".format(x / total_players) for x in total_counts]


agedemo_df = pd.DataFrame({"Total Count": total_counts,
                              "Percentage of Players": percentages},
                              index = [groupnames])
agedemo_df


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [8]:
pcount = list(groupeddemo["Purchase ID"].count())
avgpp = list(groupeddemo["Price"].mean())
totalpv = list(groupeddemo["Price"].sum())
avgppp = list(groupeddemo["Price"].sum() / groupeddemo["SN"].nunique())


panalysis_df = pd.DataFrame({"Purchase Count": pcount,
                              "Average Purchase Price": avgpp,
                              "Total Purchase Value": totalpv,
                              "Avg Total Purchase per Person": avgppp},
                              index = [groupnames])
panalysis_df["Average Purchase Price"] = panalysis_df["Average Purchase Price"].map("${:.2f}".format)
panalysis_df["Total Purchase Value"] = panalysis_df["Total Purchase Value"].map("${:.2f}".format)
panalysis_df["Avg Total Purchase per Person"] = panalysis_df["Avg Total Purchase per Person"].map("${:.2f}".format)
panalysis_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05             $1114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
usergrouped  = purchase_data.groupby("SN")
counts = usergrouped["Purchase ID"].count()
totalp = usergrouped["Price"].sum()
avgp = usergrouped["Price"].mean()

topSpenders_df = pd.DataFrame({"Purchase Count": counts, "Average Purchase Price": avgp,
                               "Total Purchase Value": totalp})
topSpenders_df = topSpenders_df.sort_values(by=['Total Purchase Value'], ascending = False)
topSpenders_df["Average Purchase Price"] = topSpenders_df["Average Purchase Price"].map("${:.2f}".format)
topSpenders_df["Total Purchase Value"] = topSpenders_df["Total Purchase Value"].map("${:.2f}".format)
topSpenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:
mostpop = purchase_data.groupby(["Item ID", "Item Name"])
purchase_count = mostpop["Purchase ID"].count()
item_price = mostpop["Price"].mean()

purchase_value = mostpop["Price"].sum()

mostpop = pd.DataFrame({"Purchase Count": purchase_count,
                        "Item Price": item_price,
                        "Total Purchase Value": purchase_value})

mostpopformat = mostpop.sort_values(by=['Purchase Count'], ascending = False)
mostpopformat["Item Price"] = mostpop["Item Price"].map("${:.2f}".format)
mostpopformat["Total Purchase Value"] = mostpop["Total Purchase Value"].map("${:.2f}".format)
mostpopformat.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [11]:
mostprofitable = mostpop.sort_values(by=['Total Purchase Value'], ascending = False)
mostprofitable["Item Price"] = mostprofitable["Item Price"].map("${:.2f}".format)
mostprofitable["Total Purchase Value"] = mostprofitable["Total Purchase Value"].map("${:.2f}".format)
mostprofitable.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# Three Observable Trends

### 1)

Based on the outputs above, one trend seen from the Heroes of Pymoli game is that males make up a majority of the player base. 84% of the players sampled seem to identify as male, which can be useful information when it comes to who to target in potential expansions, or marketing when it comes to sequels and other versions of the game. 

### 2)

Another trend seen can be from the age brackets of players, as 20-24 year olds make up about 44% of the player population. This can again be used to smartly target these players, as they are old enough maybe to want an extra challenge from the game, and yet maybe they don't have as much free time to play the game as a younger demographic. This can again help the marketing department understand who the audience for the game is. 

### 3)

Finally the last trend to mention stems from the sorting portion of the code, where Final Critic appeared to be the top most popular item, as well as the the top most profitable item. And again on both, Oathbreaker, Last Hope of the Breaking Storm, was the next most popular and profitable item. This is good for determining microtransactions for thep player, as we now have an idea of what items people would like to spend extra money on, and we can study why and how these items are so popular/profitable. Again, when a new expansion inevitably comes out, we can perhaps have new items similar to these two, in hopes of increased profit. 